packages/spl-token/src/api/resolvers.ts

```diff
@@ -0,0 +1,168 @@
+import {
+  SPLAccountBalance,
+  SPLAccountHoldings,
+  SPLTokenEvent,
+  SPLTokenInfo,
+} from '../types.js'
+import BN from 'bn.js'
+import MainDomain from '../domain/main.js'
+import { ALEPH_MINT_ADDRESS } from '../constants.js'
+
+export type TokenFilters = {
+  mint?: string
+}
+
+export type TokenEventsFilters = {
+  mint: string
+  account?: string
+  types?: string[]
+  startDate?: number
+  endDate?: number
+  limit?: number
+  skip?: number
+  reverse?: boolean
+}
+
+export type AccountHoldingsFilters = {
+  mint: string
+  account?: string
+  startDate?: number
+  endDate?: number
+  gte?: string
+  lte?: string
+}
+
+export type TokenHoldersFilters = {
+  mint: string
+  account?: string
+  limit?: number
+  skip?: number
+  reverse?: boolean
+  gte?: string
+  lte?: string
+}
+
+export class APIResolver {
+  constructor(protected domain: MainDomain) {}
+
+  async getTokens(mint?: string): Promise<SPLTokenInfo[]> {
+    return await this.filterTokens({ mint })
+  }
+
+  async getAccountHoldings({
+    mint,
+    account,
+    startDate,
+    endDate,
+    gte,
+  }: AccountHoldingsFilters): Promise<SPLAccountHoldings[]> {
+    const accountHoldings = await this.domain.getAccountHoldings(mint, {
+      account,
+      startDate,
+      endDate,
+      gte,
+    })
+
+    return accountHoldings.sort((a, b) => a.account.localeCompare(b.account))
+  }
+
+  async getTokenEvents({
+    mint,
+    account,
+    types,
+    startDate = 0,
+    endDate = Date.now(),
+    limit = 1000,
+    skip = 0,
+    reverse = true,
+  }: TokenEventsFilters): Promise<SPLTokenEvent[]> {
+    if (limit < 1 || limit > 1000)
+      throw new Error('400 Bad Request: 1 <= limit <= 1000')
+
+    const typesMap = types ? new Set(types) : undefined
+
+    const events = await this.domain.getMintEvents(mint, {
+      account,
+      startDate,
+      endDate,
+      reverse,
+      limit: !typesMap ? limit + skip : undefined,
+    })
+
+    return events
+  }
+
+  async getTokenHolders({
+    mint,
+    limit,
+    skip = 0,
+    reverse = true,
+    gte,
+    lte,
+  }: TokenHoldersFilters): Promise<SPLAccountBalance[]> {
+    const isALEPH = mint === ALEPH_MINT_ADDRESS
+
+    // @note: Default limit and gte
+    limit = limit || (isALEPH ? Number.MAX_SAFE_INTEGER : 1000)
+
+    // @note: Do not add constraints to limit arg when it is ALEPH token
+    if (!isALEPH && (limit < 1 || limit > 1000))
+      throw new Error('400 Bad Request: 1 <= limit <= 1000')
+
+    const gteBn = gte ? new BN(gte) : undefined
+    const lteBn = lte ? new BN(lte) : undefined
+
+    const result: SPLAccountBalance[] = []
+    const balances = await this.domain.getTokenHolders(mint)
+
+    let sortedBalances: SPLAccountBalance[] = []
```

Comment:
We need the list ordered by balance, and because is not possible to get the entity list ordered by this key, we need to iterate through the entire balances database to get all of them, order it, and filter by limit and skip options.